toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -234,22 +234,23 @@ class DiagnosticEmitter {
     explicit DiagnosticBuilder(
         DiagnosticEmitter<LocationT>* emitter, LocationT location,
         const Internal::DiagnosticBase<Args...>& diagnostic_base,
-        Internal::NoTypeDeduction<Args>... args)
+        llvm::SmallVector<llvm::Any> args)
         : emitter_(emitter),
-          diagnostic_({.level = diagnostic_base.Level,
-                       .message = MakeMessage(location, diagnostic_base,
-                                              std::move(args)...)}) {
+          diagnostic_(
+              {.level = diagnostic_base.Level,
+               .message = MakeMessage(emitter, location, diagnostic_base,
+                                      std::move(args))}) {
       CARBON_CHECK(diagnostic_base.Level != DiagnosticLevel::Note);
     }
 
     template <typename... Args>
-    auto MakeMessage(LocationT location,
-                     const Internal::DiagnosticBase<Args...>& diagnostic_base,
-                     Internal::NoTypeDeduction<Args>... args)
-        -> DiagnosticMessage {
+    static auto MakeMessage(
+        DiagnosticEmitter<LocationT>* emitter, LocationT location,
+        const Internal::DiagnosticBase<Args...>& diagnostic_base,
+        llvm::SmallVector<llvm::Any> args) -> DiagnosticMessage {
```

Comment:
Do you know why the old approach here stopped working? As far as I can see, the diagnostic argument types used by the driver diagnostics are `std::string` and `int64_t`, and we have other diagnostics that take a single `std::string` argument. Is the problem here the `int64_t`? What broke?

Moving the `SmallVector` when constructing a diagnostic might not be especially cheap, and in Clang it's not uncommon to produce thousands of warnings (that get filtered out at some later stage) in a single compilation, so it'd be nice to avoid these extra `SmallVector` moves if we can. But the whole diagnostics machinery is a little inefficient, so maybe it's not the low-hanging fruit here.

Reply:
The code still builds and the tests pass with the changes in this file (other than the changes to `DiagnosticLocation`) reverted, so I guess whatever issue you were encountering was solved a different way?

Reply:
```
╚╡git checkout trunk diagnostics/diagnostic_emitter.h
Updated 1 path from c7d54344d
╚╡bazel run //toolchain/source:source_buffer_test
INFO: Invocation ID: 02f46eec-7bc7-4d31-a6fa-eb7bdd300889
INFO: Build option --compilation_mode has changed, discarding analysis cache.
INFO: Analyzed target //toolchain/source:source_buffer_test (1 packages loaded, 1542 targets configured).
INFO: Found 1 target...
Target //toolchain/source:source_buffer_test up-to-date:
  bazel-bin/toolchain/source/source_buffer_test
INFO: Elapsed time: 5.368s, Critical Path: 4.60s
INFO: 4 processes: 1 internal, 3 linux-sandbox.
INFO: Build completed successfully, 4 total actions
INFO: Running command line: external/bazel_tools/tools/test/test-setup.sh toolchain/source/source_buffer_test
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //toolchain/source:source_buffer_test
-----------------------------------------------------------------------------
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from SourceBufferTest
[ RUN      ] SourceBufferTest.MissingFile
Please report issues to https://github.com/carbon-language/carbon-lang/issues and include the crash backtrace.
Stack dump:
0.	Program arguments: /usr/local/google/home/jperkins/.cache/bazel/_bazel_jperkins/85deb7d9d96f7e0e80b42618a55969d7/execroot/carbon/bazel-out/k8-fastbuild/bin/toolchain/source/source_buffer_test.runfiles/carbon/toolchain/source/source_buffer_test
```

It hangs there, I don't know why, but I think there's an issue with arg passing.

Reply:
(From offline discussion:) Looks like this is probably a miscompile bug in Clang 14.0, though we don't know the exact nature of the bug yet.

Reply:
```
╚╡clang --version
Debian clang version 14.0.6
```

As it turns out, although 14 is default, I also have 15 easy to install:

```
╚╡export CC=/usr/lib/llvm-15/bin/clang
╚╡$CC --version
Debian clang version 15.0.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-15/bin
```

(for reference, 16 is unavailable)

It looks like the issue does not repro with clang-15.

Note GitHub only has 14: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md

This is important for tests.

Reply:
Also for reference, I've tried writing a test that I thought would be equivalent, but did not reproduce the issue:

```
struct FilenameFakeDiagnosticLocationTranslator
    : DiagnosticLocationTranslator<llvm::StringRef> {
  auto GetLocation(llvm::StringRef s) -> DiagnosticLocation override {
    return {.file_name = s};
  }
};

TEST_F(DiagnosticEmitterTest, EmitLLVMErrorMessage) {
  EXPECT_CALL(consumer_, HandleDiagnostic(
                             IsDiagnostic(DiagnosticKind::TestDiagnostic,
                                          DiagnosticLevel::Error, -1, -1, _)));
  std::error_code ec(3, std::generic_category());
  llvm::ErrorOr<bool> err(ec);
  FilenameFakeDiagnosticLocationTranslator translator;
  DiagnosticEmitter<llvm::StringRef> emitter(translator, consumer_);
  llvm::StringRef filename = "x";
  if (err.getError()) {
    CARBON_DIAGNOSTIC(TestDiagnostic, Error, "Error: {0}", std::string);
    emitter.Emit(filename, TestDiagnostic, err.getError().message());
  }
}
```

Small variations on the above also did not reproduce.